### DISCUSSION TITLE
config.fish is called by every fzf preview

### DISCUSSION THREAD

**Discussion by raven2cz** [PatrickF1/fzf.fish#182] [Q&A] [answered]
## Describe the bug

I found some potential error. I expect that `config.fish` has to be called just one time when fish is launched. 

For example I "wrongly" inserted ascii art function to the last line of config.fish:

`colorscript random` - script created by DT, gitlab resource.  

I started fzf history search and the colorscript is shown instead of command preview. I started search file, and color script is shown instead of preview of file. 

Maybe it is some behavior of fish, but there has to be explanation why the lines of config.fish are call if your preview is shown. If the config.fish is called for each preview, it is very bad. 

## Steps to Reproduce

Download DT gitlab colorscript, place last row `coloscript random` to config.fish and test fzf.

## Screenshots

![fish-error](https://user-images.githubusercontent.com/20504243/123534188-8cb9ae80-d71b-11eb-936d-7c6988968699.png)

## Environment

Versions installed:

- fish, version 3.2.2
- fzf.fish: 7.1
- plugin manager: 4.3.0
- terminal: alacritty
- OS: arch

**Comment by kidonng** [comment]
This is expected. Preview spawns a new instance of fish, so if you put unrelated stuff in `config.fish`, they will be executed as well.
The correct way is to only run those stuff in interactive mode: https://github.com/fish-shell/fish-shell/pull/7796#discussion_r598227953

> If the config.fish is called for each preview, it is very bad.

It is not. fzf _has to_ create preview by running commands in a new shell. If you use e.g. zsh it will need to load `.zshrc` as well.

**Comment by PatrickF1** [comment, marked as answer]
Yup, the way fzf previews work is they spawn a new instance of $SHELL. And whenever a new shell starts up, it loads its config files.

**Comment by raven2cz** [comment]
Thanks for info.

So, it is just calling of `config.fish` for this preview instance, or some allocation of all next functions in the fish directory? Because I have a lot of functions defined in the `config.fish`. For example if I moved it to the function part, if it will be more efficient. 

I don't use interactive block, but I used

```
# Fish greetings when the terminal is shown up in interactive mode
function fish_greeting
    colorscript random
end
```

Is it correct or there is some hidden problem?

Just small question before closing this issue. The preview is using bat, is there some possibility to show images and pdf files content? Can you give me some trick to it, thanks again.

**Comment by PatrickF1** [comment]
> So, it is just calling of config.fish for this preview instance, or some allocation of all next functions in the fish directory?

fzf preview loads a new shell. Whenever a new shell starts up, it loads config.fish.

> Because I have a lot of functions defined in the config.fish. For example if I moved it to the function part, if it will be more efficient.

I'd recommend moving your functions out of config.fish and defining them as autoloaded functions.

> Is it correct or there is some hidden problem?

Seems fine to me.

> The preview is using bat, is there some possibility to show images and pdf files content?

Check out https://github.com/PatrickF1/fzf.fish/wiki/Cookbook#how-can-i-customize-file-previews-based-on-the-file-type.

**Comment by PatrickF1** [comment]
Related question: https://github.com/PatrickF1/fzf.fish/discussions/254.
